File: samples/Linux/netcore_for_Linux_x64_only/netcore1.0/TableManagerSamples/GetOffers/src/Offer.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace GetOffers
{
    class Offer
    {
        public DateTime Date
        {
            get
            {
                return mDate;
            }
            set
            {
                mDate = value;
            }
        }
        private DateTime mDate;

        public double Bid
        {
            get
            {
                return mBid;
            }
            set
            {
                mBid = value;
            }
        }
        private double mBid;

        public double Ask
        {
            get
            {
                return mAsk;
            }
            set
            {
                mAsk = value;
            }
        }
        private double mAsk;

        public string OfferID
        {
            get
            {
                return mOfferID;
            }
            set
            {
                mOfferID = value;
            }
        }
        private string mOfferID;

        public string Instrument
        {
            get
            {
                return mInstrument;
            }
            set
            {
                mInstrument = value;
            }
        }
        private string mInstrument;

        public int Precision
        {
            get
            {
                return mPrecision;
            }
            set
            {
                mPrecision = value;
            }
        }
        private int mPrecision;

        public double PipSize
        {
            get
            {
                return mPipSize;
            }
            set
            {
                mPipSize = value;
            }
        }
        private double mPipSize;

        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="sOfferID"></param>
        /// <param name="sInstrument"></param>
        /// <param name="iPrecision"></param>
        /// <param name="dPipSize"></param>
        /// <param name="date"></param>
        /// <param name="dBid"></param>
        /// <param name="dAsk"></param>
        public Offer(string sOfferID, string sInstrument, int iPrecision, double dPipSize, DateTime date, double dBid, double dAsk)
        {
            mOfferID = sOfferID;
            mInstrument = sInstrument;
            mPrecision = iPrecision;
            mPipSize = dPipSize;
            mDate = date;
            mBid = dBid;
            mAsk = dAsk;
        }
    }

    class OfferCollection : IEnumerable<Offer>
    {
        private List<Offer> mOffers;
        private Dictionary<string, Offer> mIDsAndOffers;
        private Object mSyncObj;

        /// <summary>
        /// ctor
        /// </summary>
        public OfferCollection()
        {
            mOffers = new List<Offer>();
            mIDsAndOffers = new Dictionary<string,Offer>();
            mSyncObj = new Object();
        }

        /* Add offer to collection */
        public void AddOffer(Offer offer)
        {
            lock (mSyncObj)
            {
                if (!mIDsAndOffers.ContainsKey(offer.OfferID))
                {
                    mIDsAndOffers.Add(offer.OfferID, offer);
                }
                else
                {
                    mIDsAndOffers[offer.OfferID] = offer;
                }
                mOffers.Clear();
                mOffers.AddRange(mIDsAndOffers.Values);
            }
        }

        /* Find offer by id */
        public bool FindOffer(string sOfferID, out Offer offer)
        {
            bool result = false;
            lock (mSyncObj)
            {
                offer = null;
                if (mIDsAndOffers.ContainsKey(sOfferID))
                {
                    offer = mIDsAndOffers[sOfferID];
                    result = true;
                }
            }
            return result;
        }

        /* Get offer by index */
        public Offer GetOffer(int index)
        {
            Offer offer = null;
            lock (mSyncObj)
            {
                if (index >= 0 && index < mOffers.Count)
                    offer = mOffers[index];
            }
            return offer;
        }

        /* Get number of offers */
        public int Size()
        {
            int size = 0;
            lock (mSyncObj)
            {
                size = mOffers.Count;
            }
            return size;
        }

        /* Clear collection */
        public void Clear()
        {
            lock (mSyncObj)
            {
                mOffers.Clear();
                mIDsAndOffers.Clear();
            }
        }

        #region IEnumerable<Offer> Members

        public IEnumerator<Offer> GetEnumerator()
        {
            return mOffers.GetEnumerator();
        }

        #endregion

        #region IEnumerable Members

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return mOffers.GetEnumerator();
        }

        #endregion
    }
}

File: samples/cpp_VS2005/NonTableManagerSamples/GetOffers/source/Offer.cpp
#include "stdafx.h"
#include <string>
#include <vector>
#include <map>
#include "Offer.h"

/** Constructor. */
Offer::Offer(const char *id, const char *instrument, int precision, double pipsize, DATE date, double bid, double ask)
{
    mID = id;
    mInstrument = instrument;
    mPrecision = precision;
    mPipSize = pipsize;
    mDate = date;
    mBid = bid;
    mAsk = ask;
}

/** Update bid. */
void Offer::setBid(double bid)
{
    mBid = bid;
}

/** Update ask. */
void Offer::setAsk(double ask)
{
    mAsk = ask;
}

/** Update date. */
void Offer::setDate(DATE date)
{
    mDate = date;
}

/** Get id. */
const char *Offer::getID() const
{
    return mID.c_str();
}

/** Get instrument. */
const char *Offer::getInstrument() const
{
    return mInstrument.c_str();
}

/** Get precision. */
int Offer::getPrecision() const
{
    return mPrecision;
}

/** Get pipsize. */
double Offer::getPipSize() const
{
    return mPipSize;
}

/** Get bid. */
double Offer::getBid() const
{
    return mBid;
}

/** Get ask. */
double Offer::getAsk() const
{
    return mAsk;
}

/** Get date. */
DATE Offer::getDate() const
{
    return mDate;
}


/** Constructor. */
OfferCollection::OfferCollection()
{
}

/** Destructor. */
OfferCollection::~OfferCollection()
{
    clear();
}

/** Add offer to collection. */
void OfferCollection::addOffer(Offer *offer)
{
    sample_tools::Mutex::Lock lock(m_Mutex);
    std::map<std::string, Offer *>::const_iterator iter = mIndex.find(offer->getID());
    if (iter != mIndex.end())
    {
        for (int i = 0; i < (int)mOffers.size(); i++)
        {
            if (mOffers[i] == offer)
            {
                delete mOffers[i];
                mOffers.erase(mOffers.begin() + i);
                break;
            }
        }
    }
    mOffers.push_back(offer);
    mIndex[offer->getID()] = offer;
}

/** Find offer by id. */
Offer *OfferCollection::findOffer(const char *id) const
{
    sample_tools::Mutex::Lock lock(m_Mutex);
    std::map<std::string, Offer *>::const_iterator iter = mIndex.find(id);
    if (iter == mIndex.end())
        return 0;
    else
        return iter->second;
}

/** Get number of offers. */
int OfferCollection::size() const
{
    sample_tools::Mutex::Lock lock(m_Mutex);
    return (int)mOffers.size();
}

/** Get offer by index. */
Offer *OfferCollection::get(int index) const
{
    sample_tools::Mutex::Lock lock(m_Mutex);
    return mOffers[index];
}

void OfferCollection::clear()
{
    sample_tools::Mutex::Lock lock(m_Mutex);
    int i;
    for (i = 0; i < (int)mOffers.size(); ++i)
        delete mOffers[i];
    mOffers.clear();
    mIndex.clear();
}

